### ISSUE TITLE
Go proxy doesn't boot on macOS 10.12.5

### ISSUE THREAD

**Issue by jmhodges** [bazelbuild/rules_webtesting#150]
Discussed in #147, but it seemed different enough to #148, that it should be its own ticket.

The Go proxy here doesn't boot on macOS 10.12.5 because the domain returned by `httphelpers.FQDN` returns the `os.Hostname` result which does not actually map to any IP addresses.

~~This seems to happen on my machine when the hostname (as visible in the console, too) differs from the domain talked about in the Sharing section of the macOS System Preferences.~~

It seems we're trying to boot on a public net interface, but I don't know why.

The error:

2017/06/28 19:30:32 launching HTTP server at: JeffsPersonalMBP:63762
Starting ChromeDriver 2.29.461585 (0be2cd95f834e9ee7c46bcc7cf405b483f5ae83b) on port 63757
Only local connections are allowed.
2017/06/28 19:30:33 [WebDriver proxy]: context deadline exceeded waiting for healthy: [WebDriver proxy]: error getting http://JeffsPersonalMBP:63762/healthz: Get http://JeffsPersonalMBP:63762/healthz: dial tcp: lookup JeffsPersonalMBP: no such host

**Comment by jmhodges** [comment]
(The comment about the hostnames are misleading other than the value returned by `os.Hostname` is not something you can look up. It was a bad hypothesis that failed under checks with `dig` to verify)

**Comment by joshbruning** [comment]
I recommend investigating replacing os.Hostname fallback with "localhost" fallback in the short term; I can check that that doesn't cause issues with our Linux workstations tomorrow.

**Comment by jmhodges** [comment]
Yep, that works!

**Comment by jmhodges** [comment]
Thank you!

**Comment by jmhodges** [comment]
So, here's something cool. I tried this on a different machine on my network, and the code fails. I made a ticket against the Go repository but I'm not actually sure the Go resolver is behaving incorrectly.

Here's that ticket: https://github.com/golang/go/issues/20904

Does just using `localhost` work for y'all or is this hostname lookup still needed?

**Comment by joshbruning** [comment]
Using localhost works for us in every environment we currently have (which are all Linux); for most of the history of this code in Google, we just used localhost for these addresses. We used fqdn code in test procedures for referring to the web server, though, as the browser in our VMs needed to reach it. The fqdn logic was added later when needed for another feature.

(We have a disused mac mini that we could probably experiment on...)

Thanks for filing the ticket.

I would assume the Java fqdn method works correctly on your machines?
http://docs.oracle.com/javase/8/docs/api/java/net/InetAddress.html#getCanonicalHostName--

I'm not opposed to an environment variable that forces localhost in the meantime if that unblocks you.

**Comment by jmhodges** [comment]
Hm, wasn't sure what you address or hostname for me to use to create the InetAddress. I went with a few different ones and found one that was different. Lmk if you had something else in mind.

```
import java.net.InetAddress;

public class Foo {
    public static void main(String[] args) throws Exception {
        InetAddress addr = InetAddress.getByAddress(new byte[]{127,0,0,1});
        System.out.println(addr.getCanonicalHostName());

        addr = InetAddress.getByName("localhost");
        System.out.println(addr.getCanonicalHostName());

        addr = InetAddress.getLocalHost();
        System.out.println(addr.getCanonicalHostName());
    }
}
```

outputs

```
localhost
localhost
homeportal
```

**Comment by jmhodges** [comment]
Oh! You meant `InetAddress.getByName("iMac")`, of course.

That prints out `homeportal`, too.

**Comment by joshbruning** [comment]
Sorry, forgot that wasn't a static method... Meant InetAddress.getLocalHost().getCanonicalHostName() which should result in "homeportal" per your experimentation. getCanonicalHostName would ideally return the fqdn of your machine... Looks like Go maintainers are following up.

**Comment by jmhodges** [comment]
Yeah, if it's happening in both places, maybe it's a platform bug.

btw, I made that patch for that env var but I'm not sure how to pass an env var to the go_web_test_suite target. I'd rather avoid having to put `--action_env` in my bazelrc and the env var in my bash_profile if I can because it feels a little easy to forget about.

But maybe that's the only option when using an env var?

**Comment by joshbruning** [comment]
You can either:
1. rename the environment variable to mean the opposite and make localhost use default in project (SERVE_ON_FQDN or some better name)
2. https://github.com/bazelbuild/rules_webtesting/blob/master/web/internal/web_test.sh.template - check for presence of some SERVE_ON_LOCALHOST-like environment variable; if not present, set to a truthy default.

I probably prefer option 2 just b/c we already branch the shell script template internally.

I think it's OK to use localhost by default for this project and let anyone requiring FQDN (no one yet) to opt-in as the logic is currently causing issues for mac - and I don't believe any of the VM stuff or optional features requiring fqdn for remote browsers are open sourced at this point.

**Comment by jmhodges** [comment]
I don't quite follow option 2's description. Would a different env var be set and looked at in the Go code?

**Comment by joshbruning** [comment]
Recently back from vacation... More details...

in the web_test.sh.template, was thinking:

if [[ -z "$SERVE_ON_LOCALHOST" ]]; then
  export SERVE_ON_LOCALHOST=true
fi

Then there is go/launcher/cmdhelper which would allow something like:
if cmdhelper.IsTruthyEnv("SERVE_ON_LOCALHOST") {
  return "localhost"
}

******
Also would be fine with always returning "localhost" for Mac platform for now.

Feel free to make whatever change makes this work for you by default - without having to pass any additional options or customize your blazerc (as long as there is some env var way for us to switch it off). Or feel free to request that I just make a change and I can throw something passable together.

**Comment by jmhodges** [comment]
Okay, I've not gotten any closer on this. I don't know why we couldn't use localhost everywhere?

Also, could this be reopened?

**Comment by Ubehebe** [comment]
I just ran into this issue. A web_test target I'd written suddenly started timing out with no code changes. @jmhodges, thanks for your excellent documentation here and in https://github.com/golang/go/issues/20904.

In my case, I was able to remove the offending name server from my OS's DNS configuration, and the test passed. But I'm strongly in favor of using `localhost` by default wherever possible. DNS configuration is a complex source of nondeterminism. If users of rules_webtesting want that nondeterminism, IMO it should be opt-in rather than opt-out.

@jmhodges is your fork suitable for merging? I'm reasonably good at ~~Sky~~Starlark and can help out.

**Comment by jmhodges** [comment]
It's not been updated recently, unfortunately. I didn't do anything other than the minimum required to get `FQDN` to always return `"localhost"`, though.

**Comment by jmhodges** [comment]
Also, you might want to make a new ticket since this one is closed and they don't seem to be commenting on it

**Comment by Ubehebe** [comment]
@joshbruning would you mind reopening this issue? I could file a new one, but it would refer to most of the discussion here.

In any case, I'll try to send a PR soon.

**Comment by jmhodges** [comment]
Ping

**Comment by DrMarcII** [comment]
Is this still an issue for you?
